cmd/kubeadm/app/util/patches/patches_test.go

```diff
@@ -0,0 +1,395 @@
+		expectedPatchSet *patchSet
+		data             string
+	}{
+		{
+			name:       "valid: YAML patches are separated and converted to JSON",
+			targetName: "etcd",
+			patchType:  types.StrategicMergePatchType,
+			data:       "foo: bar\n---\nfoo: baz\n",
+			expectedPatchSet: &patchSet{
+				targetName: "etcd",
+				patchType:  types.StrategicMergePatchType,
+				patches:    []string{`{"foo":"bar"}`, `{"foo":"baz"}`},
+			},
+		},
+		{
+			name:       "valid: JSON patches are separated",
+			targetName: "etcd",
+			patchType:  types.StrategicMergePatchType,
+			data:       `{"foo":"bar"}` + "\n---\n" + `{"foo":"baz"}`,
+			expectedPatchSet: &patchSet{
+				targetName: "etcd",
+				patchType:  types.StrategicMergePatchType,
+				patches:    []string{`{"foo":"bar"}`, `{"foo":"baz"}`},
+			},
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			ps, _ := createPatchSet(tc.targetName, tc.patchType, tc.data)
+			if !reflect.DeepEqual(ps, tc.expectedPatchSet) {
+				t.Fatalf("expected patch set:\n%+v\ngot:\n%+v\n", tc.expectedPatchSet, ps)
+			}
+		})
+	}
+}
+
+func TestGetPatchSetsForPathMustBeDirectory(t *testing.T) {
+	tempFile, err := ioutil.TempFile("", "test-file")
+	if err != nil {
+		t.Errorf("error creating temporary file: %v", err)
+	}
+	defer os.Remove(tempFile.Name())
+
+	_, _, _, err = getPatchSetsFromPath(tempFile.Name(), testKnownTargets)
+	if err == nil {
+		t.Fatalf("expected error for non-directory path %q", tempFile.Name())
+	}
+}
+
+func TestGetPatchSetsForPath(t *testing.T) {
+	tests := []struct {
+		name                 string
+		filesToWrite         []string
+		expectedPatchSets    []*patchSet
+		expectedPatchFiles   []string
+		expectedIgnoredFiles []string
+		expectedError        bool
+	}{
+		{
+			name:         "valid: patch files are sorted and non-patch files are ignored",
+			filesToWrite: []string{"kube-scheduler+merge.json", "kube-apiserver+json.yaml", "etcd.yaml", "foo", "bar.json"},
+			expectedPatchSets: []*patchSet{
+				{
+					targetName: "etcd",
+					patchType:  types.StrategicMergePatchType,
+				},
+				{
+					targetName: "kube-apiserver",
+					patchType:  types.JSONPatchType,
+				},
+				{
+					targetName: "kube-scheduler",
+					patchType:  types.MergePatchType,
+				},
+			},
+			expectedPatchFiles:   []string{"etcd.yaml", "kube-apiserver+json.yaml", "kube-scheduler+merge.json"},
+			expectedIgnoredFiles: []string{"bar.json", "foo"},
+		},
+		{
+			name:          "invalid: bad patch type in filename returns and error",
+			filesToWrite:  []string{"kube-scheduler+foo.json"},
+			expectedError: true,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			tempDir, err := ioutil.TempDir("", testDirPattern)
+			if err != nil {
+				t.Fatal(err)
+			}
+			defer os.RemoveAll(tempDir)
+
+			for _, file := range tc.filesToWrite {
+				filePath := filepath.Join(tempDir, file)
+				err := ioutil.WriteFile(filePath, []byte{}, 0644)
```

Comment:
If I got right this test, we are accepting empty files as a valid patch.
What about discarding empty files from the patch set with a clear log message for the users?

Reply:
messaging on empty patches makes sense.